### DISCUSSION TITLE
Help please.

### DISCUSSION THREAD

**Discussion by imarisimpson** [dnschneid/crouton#4441] [Q&A] [answered]
When I run the sudo install -Dt /usr/local/bin -m 755 ~/Downloads/crouton command it says Failed to complete chroot setup.Someone please help.

**Comment by DennisLfromGA** [comment]
That install command, if successful, has no output, it just installs crouton in your PATH so it's executable.

I believe you might be having trouble installing the chroot, that command would start with `sudo crouton ...` 
Please show us the crouton command that's failing and we may be able to help with that.

*Hope this helps,*
-**DennisLfromGA**

**Reply by imarisimpson** [reply]
I don't understand sorry sir.

**Reply by imarisimpson** [reply]
Oooo I'm sorry you mean this?sudo crouton -t unity but I also tried xfce.

**Reply by imarisimpson** [reply]
Same thing happened.

**Comment by DennisLfromGA** [comment]
Please show us the *entire* crouton command that's failing and we may be able to help with that.

Hope this helps,
-DennisLfromGA

**Reply by imarisimpson** [reply]
Ok I am confused.Am I on the right track?

**Comment by DennisLfromGA** [comment]
I'm not sure, without any further details it's hard to tell.
Please show us a **screenshot** of what you are doing and what the errors are.

*Hope this helps,*
-**DennisLfromGA**

**Reply by imarisimpson** [reply]
Ok gotcha.

**Reply by imarisimpson** [reply]
![Screenshot 2021-03-19 5 21 34 PM](https://user-images.githubusercontent.com/80975489/111843878-511b7700-88d8-11eb-9a34-7fa81de5ecb2.png)
![Screenshot 2021-03-19 5 23 43 PM](https://user-images.githubusercontent.com/80975489/111843890-54166780-88d8-11eb-839f-c10e718574d5.png)

**Reply by ghost** [reply]
Could you try a different release or target, I suggest `sudo crouton -t xfce -r sid`?
Make sure to run `sudo edit-chroot -d xenial`.

**Reply by imarisimpson** [reply]
No didn't work :/

**Comment by DennisLfromGA** [comment, marked as answer]
I tried both xenial & bionic with both unity & xfce and, after looking at the debbootstrap.log file, discovered that the root error is kernel revisions > 255, see below:

```
Selecting previously unselected package libc6:amd64.
    292 (Reading database ... 380 files and directories currently installed.)
    293 Preparing to unpack .../libc6_2.27-3ubuntu1_amd64.deb ...
    294 ERROR: Your kernel version indicates a revision number
    295 of 255 or greater.  Glibc has a number of built in
    296 assumptions that this revision number is less than 255.
    297 If you\'ve built your own kernel, please make sure that any
    298 custom version numbers are appended to the upstream
    299 kernel number with a dash or some other delimiter.
```

It was originally reported below but it affects more than just arm devices in my experience:
- crouton on arm, kernel version issue with libc updating in chroot · Issue #4429 · dnschneid/crouton
- sid installation issues on Chrome OS 90.0.4430.11 #4434

The crouton devs are aware of this issue and are looking into it.

*Hope this helps,*
-**DennisLfromGA**

**Reply by imarisimpson** [reply]
Tysm hopefully they are able to fix it.

**Reply by imarisimpson** [reply]
So do you think this can be fixed or no?

**Reply by imarisimpson** [reply]
But it's crazy because I have downloaded this so many times before and never had that problem so it's random.

**Reply by viridianknight7** [reply]
Hi, i have the same problem, and i know you said that devs are looking into a solution, but will there actually _be_ a solution? I just am concerned that this will be an abandoned issue. Hopefully, this post will be updated with a fix and I look forward to the soution.

**Comment by crowediddly** [comment]
I had the exact same issue, all available linux distros, multiple powerwashes, etc, on a 2015 Toshiba Chromebook 2. Turns out, the issue, at least on my end, was the distro channel I was on. All these errors occur on the chrome-os Developer's Channel. I finally went to the Stable channel, all issues went away, and install was simple as always.

**Reply by yoiammaac** [reply]
I am in stable though......

**Comment by jmjac** [comment]
Same issue on Chrome OS 90.0.4430.1 with kernel 4.4.257. Power washing and changing channels didn't help, but there is a temporary workaround (works for bionic and focal, xenial doesn't continue the installation after the interruption).  

Crate a custom binary  uname that  on  "uname -r" returns your kernel version x.y.z but with z lower than 255. In my case it was 4.4.254. Do <code>sudo chmod +x /path/to/custome/uname</code> to add execution permissions. Then during the installation process cancel it after "Moving bootstrap files into the chroot...". Backup the original uname with  <code>cd /usr/local/chroots/[chroot-name]/bin && sudo mv uname uname_original</code>and then copy the custom one with <code>sudo cp /path/to/custom/uname /usr/local/chroots/[chroot-name]/bin</code>. In another terminal do <code>watch -n 0.2 sudo cp /path/to/custome/uname /usr/local/chroots/[chroot-name]/bin</code> to repeat copy every 0.2s. Do <code>sudo crouton -u -n [chroot-name]</code> and continue. Later during the unpacking of coreutils your custom uname binary is going to be overwritten but the watch commad is going to take care of that. After you see "Configuring required packages..." the installation should continue without any problems and you can cancel the watch that was running in another terminal . This probably works because libc uses uname to lookup the kernel version.

After the successful installation you can replace the cusom uname with the backup of the original one with <code>cd /usr/local/chroots/[chroot-name]/bin && sudo mv uname_original uname</code>.

**Reply by svioletg** [reply]
How did you create the custom uname? I tried to follow [this](https://askubuntu.com/questions/651163/can-i-change-uname) to make one, but at `sudo mv /bin/uname /bin/uname1` it gives me `mv: cannot move '/bin/uname' to '/bin/uname1': Read-only file system`

**Reply by jmjac** [reply]
I used this https://pastebin.com/E3FrbpZE. Just fill it in with your output from uname in crosh shell and compile it with <code>go build uname.go</code> in crostini or on another device where you have go installed. You don't want to replace uname in /usr/bin, that's the chrome OS's uname. You want to replace the chroot's uname that will be located in /usr/local/chroots/[chroot-name]/usr/bin. This will be created after "Moving bootstrap files into the chroot" during the crouton installation. 

Btw. I'm not sure if there are no negative side effects when it comes to chroots stability after installing the chroot that way.

**Reply by jmjac** [reply]
You can install go in crostini and compiled uname.go there with <code>go build uname.go</code>, later move uname to Downloads in chrome os for easier path when copying.

**Reply by siddhantsid** [reply]
Bro  i dont have crostini support is there is any other way 

This is the output of uname -a 
Linux localhost 4.4.257-19935-g0ab0a71a40fd #1 SMP PREEMPT Mon Apr 19 20:04:52 PDT 2021 x86_64 Intel(R) Celeron(R) CPU N2840 @ 2.16GHz GenuineIntel GNU/Linux
chronos@localhost /mnt/stateful_partitio

**Reply by jmjac** [reply]
You could probably install chromebrew and use it to install go, but I have no experience with chromebrew at all. The uname that you want to create is really simple so you can do it in any language that creates a binary executable. It doesn't have to be go, I used it because it was the quickest to write for me.

**Reply by siddhantsid** [reply]
Thanks jmjackowiak
I successfully installed xenial

**Reply by yoiammaac** [reply]
can you give me tutorial or something with this method jmjac

**Reply by morphisjustfun** [reply]
You deserve a medal. BTW if you cant compile in another machine just use replit. It allows you to compile online and then you can download it.

**Comment by dinki** [comment]
Unfortunately I've ended up in the same place as some of these good folks.  Can anyone tell me if a working solution has been found?  I'm having the exact same problem as shown in the screenshots here:
https://github.com/dnschneid/crouton/discussions/4441#discussioncomment-505094

Any help is most appreciated.

**Reply by siddhantsid** [reply]
yup try this method 
https://github.com/dnschneid/crouton/discussions/4441#discussioncomment-700695

**Reply by dinki** [reply]
Thanks @siddhantsid .  Can I build this on any Linux box for use on a HP Chromebook 11 G4?  I've got an old Debian box and an Ubuntu VM that I can install go on but not sure if the binaries will run on the Chromebook.

**Reply by siddhantsid** [reply]
just install go directly on your chromeos
https://github.com/golang/go/wiki/ChromeOS

**Reply by jmjac** [reply]
@dinki If you have go installed on different device you can compile it for chromebook with <code>GOOS=linux GOARCH=[your chromebook's arch] go build</code> and then just rename the resulting src file to uname. Use <code>go tool dist list</code> for the list of available os/arch pairs.

**Reply by dinki** [reply]
@siddhantsid 

Thanks for this.  I did manage to build but now having trouble with the install.  I killed it at the "Moving bootstrap files.." part and then did this:

```
chronos@localhost /usr/local/chroots/bionic/bin $ sudo mv uname uname_original
chronos@localhost /usr/local/chroots/bionic/bin $ sudo cp /home/chronos/user/Downloads/uname .
chronos@localhost /usr/local/chroots/bionic/bin $ sudo enter-chroot
Entering /mnt/stateful_partition/crouton/chroots/bionic...
A chroot setup script still exists inside the chroot.
The chroot may not be fully set up.
However, it appears the setup script is invalid.
Removing the chroot setup script. You may want to update your chroot again.
/mnt/stateful_partition/crouton/chroots/bionic doesn't appear to be a valid chroot.
Unmounting /mnt/stateful_partition/crouton/chroots/bionic...
```

Any idea on what I'm doing wrong?  Thanks to you all for the help.  I think I'm close.

**Reply by siddhantsid** [reply]
instead of typing sudo enter-chroot
type this
sudo crouton -u -n bionic  (Note: type this command in new shell close the old one )

**Reply by dinki** [reply]
Thanks for this.  Still no joy:

```
I: Unpacking tar...
I: Unpacking util-linux...
I: Unpacking zlib1g:amd64...
W: Failure while unpacking required packages.  This will be attempted up to five times.
W: See //debootstrap/debootstrap.log for details (possibly the package /var/cache/apt/archives/libc6_2.27-3ubuntu1_amd64.deb is at fault)
Failed to complete chroot setup.
Unmounting /mnt/stateful_partition/crouton/chroots/bionic...

```

**Reply by jmjac** [reply]
@siddhantsid It looks like <code>sudo crouton -u -n [chroot-name]</code> works more often then enter-chroot. I updated my post to reflect that.

@dinki Sometimes it's just timing.  Try repeating the cp command in another terminal until you <code>I: Configuring required packages...</code>.

**Reply by dinki** [reply]
@jmjac  Thanks.  Can you elaborate on the timing part.  Am I to try and copy while crouton is trying to install in another terminal?  Not understanding.

**Reply by siddhantsid** [reply]
@dinki  you have not read the instructions by @jmjac  nicely < 
You have to replace it with your own again. It's a pretty short time window to do it, so I recommend quickly repeating sudo cp /path/to/custom/uname /usr/local/chroots/[chroot-name]/bin in another terminal window or setting up a script to do it for you. After you see "Configuring required packages..." >

**Reply by jmjac** [reply]
Yeah. When crouton is doing it's work after <code>sudo crouton -u -n [chroot-name]</code> you have another terminal open where you do <code>watch -n 0.2 sudo cp /path/to/custome/uname .</code> to repeat the copy command every 0.2s. If you get <code>su: pam_start: error 26</code> you got the interruption timing wrong so delete-chroot and start again. It's a lot of trial and error to learn how to get it right consistently.

**Reply by jmjac** [reply]
Also do chmod +x on the custome uname before you copy it.

**Reply by dinki** [reply]
I thought I was coming back to claim victory using the tips you guys provided, but alas, no.  It did get a lot farther but I ended up here:

```
Setting up libdrm-dev:amd64 (2.4.101-2~18.04.1) ...
Processing triggers for libc-bin (2.27-3ubuntu1.4) ...
libdrm-dev set to automatically installed.
Compiling /usr/local/lib/croutonfreon.so...
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package 'xserver-xorg-video-all' is not installed, so not removed
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 xorg : Depends: xserver-xorg (>= 1:7.7+19ubuntu7.1)
 xserver-xorg-video-intel : Depends: xorg-video-abi-23
                            Depends: xserver-xorg-core (>= 2:1.18.99.901)
E: Unable to correct problems, you have held broken packages.
Failed to complete chroot setup.
Unmounting /mnt/stateful_partition/crouton/chroots/bionic...
Sending SIGTERM to processes under /mnt/stateful_partition/crouton/chroots/bionic...
```

**Reply by jmjac** [reply]
@dinki That's a bionic problem. Try this https://github.com/dnschneid/crouton/wiki/3-Steps-to-Install-Bionic-Beaver
